File: ts/instructions/replaceNote.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ReplaceNoteAccounts {
  signer: PublicKey
  /** The global root */
  root: PublicKey
  /** The tree */
  tree: PublicKey
  /** The node the note will be attached to */
  node: PublicKey
  /** The new note */
  note: PublicKey
  /** The new note */
  weakNote: PublicKey
  /** Common Solana programs */
  systemProgram: PublicKey
}

export function replaceNote(accounts: ReplaceNoteAccounts) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.root, isSigner: false, isWritable: false },
    { pubkey: accounts.tree, isSigner: false, isWritable: false },
    { pubkey: accounts.node, isSigner: false, isWritable: true },
    { pubkey: accounts.note, isSigner: false, isWritable: true },
    { pubkey: accounts.weakNote, isSigner: false, isWritable: false },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([57, 50, 234, 92, 148, 156, 171, 166])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/dippies_index_protocol/src/instructions/replace_note.rs
use anchor_lang::prelude::*;

use crate::errors::TreeDeaErrors;
use crate::seeds::{NODE_SEED, NOTE_SEED, ROOT_SEED, TREE_SEED};
use crate::state::{Node, Note, Root, Tree, MAX_NOTES_PER_NODE};

pub fn replace_note(ctx: Context<ReplaceNote>) -> Result<()> {
    msg!("Replacing a note");

    let node = &mut ctx.accounts.node;
    let note = &mut ctx.accounts.note;

    node.stake -= ctx.accounts.weak_note.stake;
    node.stake += note.stake;

    let position = node
        .notes
        .iter()
        .position(|n| n == &ctx.accounts.weak_note.key())
        .unwrap();
    node.notes[position] = note.key();

    Ok(())
}

#[derive(Accounts)]
pub struct ReplaceNote<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The global root
    #[account(
        seeds = [
            ROOT_SEED.as_bytes(),
            &root.id.to_bytes(),
        ],
        bump,
    )]
    pub root: Account<'info, Root>,

    /// The tree
    #[account(
        seeds = [
            TREE_SEED.as_bytes(),
            &root.key().to_bytes(),
            &tree.title.as_ref(),
        ],
        bump,
    )]
    pub tree: Account<'info, Tree>,

    /// The node the note will be attached to
    #[account(
        mut,
        seeds = [
            NODE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &node.parent.key().to_bytes(),
            &node.tags.last().unwrap().as_ref(),
        ],
        bump,
        constraint = node.notes.contains(&weak_note.key()) @ TreeDeaErrors::NotOnNode,
        constraint = !node.notes.contains(&note.key()) @ TreeDeaErrors::AlreadyOnNode,
        constraint = node.notes.len() >= MAX_NOTES_PER_NODE @ TreeDeaErrors::NodeNotFull,
    )]
    pub node: Account<'info, Node>,

    /// The new note
    #[account(
        mut,
        seeds = [
            NOTE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &note.id.to_bytes()
        ],
        bump,
        constraint = note.parent == node.key() @ TreeDeaErrors::NotChildNote,
        constraint = note.stake > weak_note.stake @ TreeDeaErrors::NotEnoughStake,
    )]
    pub note: Account<'info, Note>,

    /// The new note
    #[account(
        seeds = [
            NOTE_SEED.as_bytes(),
            &tree.key().to_bytes(),
            &weak_note.id.to_bytes()
        ],
        bump,
        constraint = weak_note.parent == node.key() @ TreeDeaErrors::NotChildNote,
    )]
    pub weak_note: Account<'info, Note>,

    /// Common Solana programs
    pub system_program: Program<'info, System>,
}
